### PR TITLE
Add ability to run a test residing on a given line

### PR DESCRIPTION
**Pull request by qti3e**
:unicorn: Fixes: #2138 

There are just a few edges case that I found it useful to have your comments on :)

- Currently, negative numbers are just ignored, maybe throw an error or something?
- What should happen if the `:...` is actually part of the file name, (`test.js:5` a valid filename on Linux at least)!?
- When running in the watch mode, the line number might change and results in different tests to be executed.
- Due to the implementation using line numbers that are higher than the actual number of lines in the file will result in the last test being executed, it would be nice to have a way to get line numbers in the file without the overhead of reading the file from disk.

Also, there is definitely room for tests to be improved, I couldn't actually get very useful information from `runStatus.stats` for what I had in mind, (name of executed tests) if the current test is not enough I'd be happy to hear about a way to get that information from the API.

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Referenced issues

This pull request has been submitted to:
- [#2138: Add ability to run a test residing on a given line](https://issuehunt.io/repos/26820798/issues/2138)
---

IssueHunt has been backed by the following sponsors. [Become a sponsor](https://issuehunt.io/membership/members)
</details>

### CODE REVIEW

**Comment by sindresorhus** [comment]
Negative numbers should throw a user-friendly error.

**Comment by sindresorhus** [comment]
While `test.js:5` might be a valid file name on Linux, it’s not something anyone would actually use. It’s not a concern.

**Comment by sindresorhus** [comment]
The range syntax should work for the `files` setting in the AVA config too. I don’t see that documented.

**Comment by sindresorhus** [comment]
I’m not sure it makes sense at all to allow the range syntax when using `watch`.

**Comment by qti3e** [comment]
@sindresorhus I guess everything works as expected now, I moved the whole parser logic to a new file, and also added the support for ranges in `glob` as well as in the `files` argument in `API.run()` so that this range syntax thing works everywhere the same way (configs, CLI, etc.)

That being said now we can expect a glob pattern such as `test-*.js:5` to work just fine.

There is only one option added to the `runner` named `lines`, which determines which lines are expected to be executed (an empty array means every line).

There is no negative number allowed in a range string, which means `test.js:-4` is treated as a file with the name `test.js:-4` rather than line 4 in `test.js` and well that leads to a file not found status, which I believe is best to be addressed as part of #2158 (which I'm going to take right after this PR if that's ok), and I'd rather avoid doing this error handling as part of this PR to keep it as simple as it is and keep it clean.

/c.c @novemberborn

File: lib/runner.js [hunk 1]
```diff
@@ -52,11 +58,45 @@ class Runner extends Emittery {
...
 			}
 
+			if (this.lines.length > 0) {
+				this.runOnlyExclusive = true;
+				let callsiteLineNumber;
+
+				for (const jsCallsite of callsites()) {
+					const callsite = sourceMapSupport.wrapCallSite(jsCallsite);
+					const fileName = callsite.getFileName();
+					if (fileName === this.file) {
```
Comment (novemberborn) [inline lib/runner.js]: I imagine this will break when using AVA with pre-compiled test files. The selected file may be `build/test.js` but `source-map-support` may translate the callsite to `src/test.js`. You should check the files before the source map is applied.

This also needs a comment to explain why we need to check the files in the first place (because the `test()` may be invoked in a helper file — which is its own edge case that I wouldn't know how to deal with).

File: lib/runner.js [hunk 2]
```diff
@@ -52,11 +58,45 @@ class Runner extends Emittery {
...
+						break;
+					}
+				}
+
+				if (lastMetadata) {
+					// `callsiteLineNumber` is the beginning of the current test, -1 to
+					// get the last line of the previous test.
+					const range = [lastCallsiteLineNumber, callsiteLineNumber - 1];
+					const includedInRange = this.lines.some(n => n >= range[0] && n <= range[1]);
+					lastMetadata.exclusive = this.match.length === 0 ?
```
Comment (novemberborn) [inline lib/runner.js]: `exclusive` will be `true` if the test is declared using `.only()`. But, crucially, tests can also be skipped.

I think the ranges should narrow down from the initial test selection. If there are exclusive tests, then only those that match the ranges should be run. If there are no exclusive tests, then skipped tests should never be run.

File: lib/runner.js [hunk 3]
```diff
@@ -52,11 +58,45 @@ class Runner extends Emittery {
...
+						includedInRange :
+						includedInRange && lastMetadata.exclusive;
+				}
+
+				lastCallsiteLineNumber = callsiteLineNumber;
+				lastMetadata = metadata;
```
Comment (novemberborn) [inline lib/runner.js]: Rather than tracking the last metedata object and callsite, perhaps you could add the callsite number to the metadata, and then adjust the test selection once all tests have been declared?

File: lib/runner.js [hunk 4]
```diff
@@ -52,11 +58,45 @@ class Runner extends Emittery {
...
+					const fileName = callsite.getFileName();
+					if (fileName === this.file) {
+						callsiteLineNumber = callsite.getLineNumber();
+						break;
+					}
+				}
+
+				if (lastMetadata) {
+					// `callsiteLineNumber` is the beginning of the current test, -1 to
+					// get the last line of the previous test.
```
Comment (novemberborn) [inline lib/runner.js]: This does include whitespace, but I think that's OK for now.

File: lib/range.js [hunk 5]
```diff
@@ -0,0 +1,66 @@
+'use strict';
+const rangeParser = require('parse-numeric-range');
+const union = require('array-union');
+
+function parseRange(files) {
+	const any = new Set();
```
[the remark this answers is not in the document]
Comment (qti3e) [reply lib/range.js]: @nov parseRange should work on both ranged paths and normal paths, imagine we have an input like `['a.js', 'a.js:4']` in this case the `a.js` path has a higher priority and should cancel out `a.js`, this set is used to store file paths that do not have a range syntax in them.

File: lib/range.js [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+			continue;
+		}
+
+		if (rangeStr === '') {
+			any.add(actualFilename);
+			ranges[actualFilename] = [];
+			continue;
+		}
+
+		const lines = rangeParser.parse(rangeStr);
```
[the remark this answers is not in the document]
Comment (qti3e) [reply lib/range.js]: Because of how this implementation handles globs I needed to have the range part as a string, and of course, we need another function to parse that part.

File: lib/range.js [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+	};
+}
+
+exports.parseRange = parseRange;
+
+const rangeRegExp = /^(\d+(-\d+)?,?)+$/;
+const splitRangeStrCache = new Map();
+
+function splitRangeStr(file) {
+	if (splitRangeStrCache.has(file)) {
```
[the remark this answers is not in the document]
Comment (qti3e) [reply lib/range.js]: It is called many times with the same input in different places, so I thought it's better to cache the result.

File: lib/range.js [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+		return [...splitRangeStrCache.get(file)];
+	}
+
+	const colonIndex = file.lastIndexOf(':');
+	const mightHaveRange = colonIndex > -1 && file[colonIndex + 1] !== '\\';
```
[the remark this answers is not in the document]
Comment (qti3e) [reply lib/range.js]: That regex I'm using might be a bit heavy so I use a fast and dirty check first and store the result in `mightHaveRange`, also there is no need for a test if you remove this check (and the regex), most of the tests will fail on windows.

File: lib/runner.js (hunk 1 above)
Comment (qti3e) [reply lib/runner.js]: We can't check the file before the source map is applied after the translation line numbers are all changed, (`a.js:5` might end up running line 6.)

File: lib/runner.js (hunk 4 above)
Comment (qti3e) [reply lib/runner.js]: I don't think there is a way to handle this case unless you parse the source file to it's AST and then extract when the function call actually ends (this information is not in the callsite data.)

**Comment by qti3e** [comment]
@novemberborn Please let me know your final decision about dropping ranged globs.

**Comment by sindresorhus** [comment]
> I don't see how this is useful. If you want to always run part of a test file, make a new test file. Our globbing logic is still really complex and I don't think we should add more complexity to it.

I agree. I didn't think that through when first commenting.

> Having reviewed the PR it definitely looks easier to implement and maintain if we only support this syntax on the CLI.

👍

**Comment by sindresorhus** [comment]
> There is no negative number allowed in a range string, which means test.js:-4 is treated as a file with the name test.js:-4 rather than line -4 in test.js and well that leads to a file not found status

I think that should be handled here. It's clearly not a valid filename, so we should give the user a helpful and relevant error message that it's an invalid range.

File: lib/range.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (novemberborn) [reply lib/range.js]: I see. Should that maybe print an error instead? Or perhaps we could run only the selected lines and print a warning "ignoring other tests from a.js since you selected lines" (well not quite that, but you get the idea).

File: lib/range.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (novemberborn) [reply lib/range.js]: I'm not too fussed about performance overheads, presumably you only run this with a handful of ranges so it'll be really hard to tell what's faster or slower.

> here is no need for a test if you remove this check (and the regex), most of the tests will fail on windows.

Are you saying the test is not necessary, or it is because of Windows?

File: lib/range.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (novemberborn) [reply lib/range.js]: I doubt that's necessary.

File: lib/range.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (novemberborn) [reply lib/range.js]: Right, so this will be a lot simpler without globs 😄

File: lib/runner.js (hunk 1 above)
Comment (novemberborn) [reply lib/runner.js]: > We can't check the file before the source map is applied after the translation line numbers are all changed, (`a.js:5` might end up running line 6.)

Yes I get that. My point though is that applying the source map may change the `fileName` of the call site.

File: lib/range.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (qti3e) [reply lib/range.js]: you can look at this from when this check was not there https://travis-ci.org/avajs/ava/builds/556309917

**Comment by qti3e** [comment]
@novemberborn PTAL, also please restart that Travis task.

File: lib/range.js [hunk 9]
```diff
@@ -0,0 +1,53 @@
+'use strict';
+const rangeParser = require('parse-numeric-range');
+
+function parseFileSelections(files) {
+	const ranges = new Map();
+	const ignored = [];
+	const filenames = [];
```
Comment (novemberborn) [inline lib/range.js]: Let's make this a set, too.

File: lib/range.js [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+
+	for (const file of files) {
+		const [actualFilename, lines] = parseFileSelection(file);
```
Comment (novemberborn) [inline lib/range.js]: With `filenames` a set, you can add the `actualFilename` to it here, regardless of any other condition.

File: lib/range.js [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+
+		if (!ranges.has(actualFilename)) {
+			ranges.set(actualFilename, []);
+			filenames.push(actualFilename);
+		}
+
+		if (lines.length === 0 && ranges.get(actualFilename).length > 0) {
```
Comment (novemberborn) [inline lib/range.js]: Let's move this up, and change the condition to `lines.length === 0 && ranges.has(actualFilename)`.

File: lib/range.js [hunk 12]
```diff
@@ -0,0 +1,53 @@
+'use strict';
+const rangeParser = require('parse-numeric-range');
+
+function parseFileSelections(files) {
+	const ranges = new Map();
+	const ignored = [];
```
Comment (novemberborn) [inline lib/range.js]: This should be a set, so you don't need to dedupe it later.

File: lib/range.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (novemberborn) [reply lib/range.js]: Right, so let's add a comment explaining this handles file paths on Windows that do not include ranges.

File: lib/range.js [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+
+exports.parseFileSelections = parseFileSelections;
+
+const rangeRegExp = /^(\d+(-\d+)?,?)+$/;
+
+function parseFileSelection(file) {
+	const colonIndex = file.lastIndexOf(':');
+	const mightHaveRange = colonIndex > -1 && file[colonIndex + 1] !== '\\';
+	const rangeStr = file.slice(colonIndex + 1);
+	const actualFilename = file.slice(0, colonIndex);
```
Comment (novemberborn) [inline lib/range.js]: This isn't really a filename though, it's still a file path. Let's rename this throughout.

File: lib/range.js [hunk 14]
```diff
@@ -0,0 +1,53 @@
+'use strict';
+const rangeParser = require('parse-numeric-range');
+
+function parseFileSelections(files) {
+	const ranges = new Map();
```
Comment (novemberborn) [inline lib/range.js]: `ranges` no longer holds ranges. It holds line numbers. Shall we rename this?

File: lib/cli.js [hunk 15]
```diff
@@ -280,7 +282,14 @@ exports.run = async () => { // eslint-disable-line complexity
 		});
 		watcher.observeStdin(process.stdin);
 	} else {
-		const runStatus = await api.run(files);
+		const {filenames, ranges, ignored} = rangeParser.parseFileSelections(files);
+		if (ignored.length > 0) {
+			for (const file of ignored) {
```
Comment (novemberborn) [inline lib/cli.js]: No need to add the length check.

File: lib/cli.js [hunk 16]
```diff
@@ -280,7 +282,14 @@ exports.run = async () => { // eslint-disable-line complexity
...
+				console.error(`Ignoring other tests from ${file} since you selected lines.`);
```
Comment (novemberborn) [inline lib/cli.js]: We can't write to the console from here, it messes with the reporters. I think for now we can use `exit()` here and explain that if you select lines, you shouldn't then also specify the same file without a line selection, and then print the offending files.

File: lib/cli.js [hunk 17]
```diff
@@ -280,7 +282,14 @@ exports.run = async () => { // eslint-disable-line complexity
 		});
 		watcher.observeStdin(process.stdin);
 	} else {
-		const runStatus = await api.run(files);
+		const {filenames, ranges, ignored} = rangeParser.parseFileSelections(files);
```
Comment (novemberborn) [inline lib/cli.js]: What do you think about executing this when `cli.input` is processed? Then in the `conf.watch` branch we can exit with an error if lines are selected.

File: lib/runner.js [hunk 18]
```diff
@@ -52,11 +56,29 @@ class Runner extends Emittery {
...
 				});
 			}
 
+			if (this.lines.length > 0) {
+				let callsiteLineNumber;
+
+				for (const jsCallsite of callsites()) {
+					const callsite = sourceMapSupport.wrapCallSite(jsCallsite);
+					const fileName = callsite.getFileName();
+					if (jsCallsite === this.file || fileName === this.file) {
```
Comment (novemberborn) [inline lib/runner.js]: Did you mean `jsCallsite.getFileName()`?

I don't think this needs to be compared to the source-map converted filename. You can't pass those on the CLI since you need to provide the file that AVA can actually load.

File: lib/runner.js [hunk 19]
```diff
@@ -52,11 +56,29 @@ class Runner extends Emittery {
...
+					this.callsiteLineNumbers.sort((a, b) => a - b);
 					hasStarted = true;
 					this.start();
 				});
 			}
 
+			if (this.lines.length > 0) {
+				let callsiteLineNumber;
+
+				for (const jsCallsite of callsites()) {
```
Comment (novemberborn) [inline lib/runner.js]: Let's rename `jsCallsite` to `callsite`.

File: lib/runner.js [hunk 20]
```diff
@@ -52,11 +56,29 @@ class Runner extends Emittery {
...
+					const callsite = sourceMapSupport.wrapCallSite(jsCallsite);
```
Comment (novemberborn) [inline lib/runner.js]: And rename this to `wrapped`.

File: lib/runner.js [hunk 21]
```diff
@@ -52,11 +56,29 @@ class Runner extends Emittery {
 			if (!scheduledStart) {
 				scheduledStart = true;
 				process.nextTick(() => {
+					this.callsiteLineNumbers.push(Infinity);
+					this.callsiteLineNumbers.sort((a, b) => a - b);
```
Comment (novemberborn) [inline lib/runner.js]: Let's do this work inside the `start()` method instead.

We should deduplicate the line numbers as well. It's an edge case but you could specify two tests on the same line.

File: lib/api.js [hunk 22]
```diff
@@ -51,7 +51,12 @@ class Api extends Emittery {
 	}
 
 	async run(files = [], runtimeOptions = {}) {
-		files = files.map(file => path.resolve(this.options.resolveTestsFrom, file));
+		const ranges = new Map(runtimeOptions.ranges || []);
+		files = files.map(file => {
+			const filename = path.resolve(this.options.resolveTestsFrom, file);
+			ranges.set(filename, ranges.get(file));
+			return filename;
+		});
```
Comment (novemberborn) [inline lib/api.js]: Why not do this processing in the CLI instead?

**Comment by novemberborn** [comment]
Closing due to inactivity. @qti3e I do hope you'll find the time to come back to this!
